config/metric.go

```diff
@@ -4,15 +4,257 @@
+		}
+		return sdkmetric.NewPeriodicReader(exp, opts...), nil
+	}
+	return nil, errors.New("no valid metric exporter")
+}
+
+func otlpHTTPMetricExporter(ctx context.Context, otlpConfig *OTLPMetric) (sdkmetric.Exporter, error) {
+	opts := []otlpmetrichttp.Option{}
+
+	if len(otlpConfig.Endpoint) > 0 {
+		u, err := url.ParseRequestURI(otlpConfig.Endpoint)
+		if err != nil {
+			return nil, err
+		}
+		opts = append(opts, otlpmetrichttp.WithEndpoint(u.Host))
+
+		if u.Scheme == "http" {
+			opts = append(opts, otlpmetrichttp.WithInsecure())
+		}
+		if len(u.Path) > 0 {
+			opts = append(opts, otlpmetrichttp.WithURLPath(u.Path))
+		}
+	}
+	if otlpConfig.Compression != nil {
+		switch *otlpConfig.Compression {
+		case compressionGzip:
+			opts = append(opts, otlpmetrichttp.WithCompression(otlpmetrichttp.GzipCompression))
+		case compressionNone:
+			opts = append(opts, otlpmetrichttp.WithCompression(otlpmetrichttp.NoCompression))
+		default:
+			return nil, fmt.Errorf("unsupported compression %q", *otlpConfig.Compression)
+		}
+	}
+	if otlpConfig.Timeout != nil {
+		opts = append(opts, otlpmetrichttp.WithTimeout(time.Millisecond*time.Duration(*otlpConfig.Timeout)))
 	}
-	mp := sdkmetric.NewMeterProvider()
-	return mp, mp.Shutdown
+	if len(otlpConfig.Headers) > 0 {
+		opts = append(opts, otlpmetrichttp.WithHeaders(otlpConfig.Headers))
+	}
+
+	return otlpmetrichttp.New(ctx, opts...)
+}
+
+func otlpGRPCMetricExporter(ctx context.Context, otlpConfig *OTLPMetric) (sdkmetric.Exporter, error) {
+	opts := []otlpmetricgrpc.Option{}
+
+	if len(otlpConfig.Endpoint) > 0 {
+		u, err := url.ParseRequestURI(otlpConfig.Endpoint)
+		if err != nil {
+			return nil, err
+		}
+		// ParseRequestURI leaves the Host field empty when no
+		// scheme is specified (i.e. localhost:4317). This check is
+		// here to support the case where a user may not specify a
+		// scheme. The code does its best effort here by using
+		// otlpConfig.Endpoint as-is in that case
+		if u.Host != "" {
+			opts = append(opts, otlpmetricgrpc.WithEndpoint(u.Host))
+		} else {
+			opts = append(opts, otlpmetricgrpc.WithEndpoint(otlpConfig.Endpoint))
+		}
+		if u.Scheme == "http" {
+			opts = append(opts, otlpmetricgrpc.WithInsecure())
+		}
+	}
+
+	if otlpConfig.Compression != nil {
+		switch *otlpConfig.Compression {
+		case compressionGzip:
+			opts = append(opts, otlpmetricgrpc.WithCompressor(*otlpConfig.Compression))
+		case compressionNone:
+			// none requires no options
+		default:
+			return nil, fmt.Errorf("unsupported compression %q", *otlpConfig.Compression)
+		}
+	}
+	if otlpConfig.Timeout != nil && *otlpConfig.Timeout > 0 {
+		opts = append(opts, otlpmetricgrpc.WithTimeout(time.Millisecond*time.Duration(*otlpConfig.Timeout)))
+	}
+	if len(otlpConfig.Headers) > 0 {
+		opts = append(opts, otlpmetricgrpc.WithHeaders(otlpConfig.Headers))
+	}
+
+	return otlpmetricgrpc.New(ctx, opts...)
+}
+
+func prometheusReader(ctx context.Context, prometheusConfig *Prometheus) (sdkmetric.Reader, error) {
```

Comment:
I would just like to confirm this, not that it is a blocker or not.

Is there a way to use the config mechanism and add a route to an existing HTTP server?  Or will this always be a separate prometheus instance and HTTP server?

Reply:
it's a good question, there's only host and port defined as parameters in the configuration today

Reply:
This is tied with the prom config.  If there isn't a host, we could instead use the [Default Prometheus instance](https://pkg.go.dev/github.com/prometheus/client_golang/prometheus#pkg-variables).  This would let users start the server on their own.

Reply:
I talked with Codeboten offline, and I think this is an acceptable approach.
We will start by requiring it initially; this lets us get this going quicker.  If there is a demand for using the already configured Default Prometheus then we can relax this requirement.